x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeSourceHandler.java

```diff
@@ -24,24 +24,55 @@
 /**
  * An {@link ExchangeSourceHandler} asynchronously fetches pages and status from multiple {@link RemoteSink}s
  * and feeds them to its {@link ExchangeSource}, which are created using the {@link #createExchangeSource()}) method.
- * {@link RemoteSink}s are added using the {@link #addRemoteSink(RemoteSink, int)}) method.
+ * {@link RemoteSink}s are added using the {@link #addRemoteSink(RemoteSink, int, ActionListener)}) method.
  *
  * @see #createExchangeSource()
- * @see #addRemoteSink(RemoteSink, int)
+ * @see #addRemoteSink(RemoteSink, int, ActionListener)
  */
 public final class ExchangeSourceHandler {
     private final ExchangeBuffer buffer;
     private final Executor fetchExecutor;
 
     private final PendingInstances outstandingSinks;
     private final PendingInstances outstandingSources;
-    private final FailureCollector failure = new FailureCollector();
+    private final boolean failFast;
+    private final FailureCollector failure;
 
-    public ExchangeSourceHandler(int maxBufferSize, Executor fetchExecutor) {
+    /**
+     * Creates a new ExchangeSourceHandler.
+     *
+     * @param maxBufferSize      the maximum size of the exchange buffer. A larger buffer reduces ``pauses`` but uses more memory,
+     *                           which could otherwise be allocated for other purposes.
+     * @param fetchExecutor      the executor used to fetch pages.
+     * @param failFast           whether this exchange source should stop fetching pages immediately when failures occur.
+     * @param completionListener a listener that will be notified when the exchange source handler fails or completes
+     */
+    public ExchangeSourceHandler(int maxBufferSize, Executor fetchExecutor, boolean failFast, ActionListener<Void> completionListener) {
         this.buffer = new ExchangeBuffer(maxBufferSize);
+        this.failFast = failFast;
         this.fetchExecutor = fetchExecutor;
         this.outstandingSinks = new PendingInstances(() -> buffer.finish(false));
         this.outstandingSources = new PendingInstances(() -> buffer.finish(true));
+        this.failure = new FailureCollector();
+        buffer.addCompletionListener(ActionListener.running(() -> {
+            final ActionListener<Void> listener = ActionListener.assertAtLeastOnce(completionListener).delegateFailure((l, unused) -> {
```

Comment:
I'm not sure I understand here - why are we ignoring the failure from `completionListener`?

Reply:
exchangeBuffer should never throw exceptions; here, we are throwing a null value, not an actual failure.